### PR TITLE
reduce regex op code time

### PR DESCRIPTION
**Pull request by ViktorHofer**
While stepping through the Regex code and analyzing the different steps to find the reason for https://github.com/dotnet/corefx/issues/26787 I noticed that the regex op code generation runs twice. a code comment also suggested that this could be simplified by using a growing data structure instead of a fixed array.

Benchmark results (based on Dans xunit benchmarks, converted to BenchmarkDotNet):

**[EDIT]: see benchmark below**

### CODE REVIEW

**Comment by danmosemsft** [comment]
Do you happen to have a ZIP file from PerfView from the perf test? Just would be interseting to look at. There are innumerable possible places to improve. Eg., should we construct Boyer Moore tables when the prefix or text are very small.

**Comment by danmosemsft** [comment]
We should probably add another perf test to exercise RegexOptions.Compiled, and one to exercise caching.

cc @lyaGrebnov and @artkpv who have expressed interest in regex perf recently. Perhaps we could collaborate on creating more perf tests and improving their results.

**Comment by danmosemsft** [comment]
> converted to BenchmarkDotNet

Perhaps it would make sense to share how to do this and how to run them, if it's easier than xunit.perf.

**Comment by danmosemsft** [comment]
That should have been @IlyaGrebnov above.

**Comment by danmosemsft** [comment]
@dotnet-bot test OSX10.12 x64 debug build (system hung

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 1]
```diff
@@ -56,7 +53,7 @@ internal static RegexCode Write(RegexTree t)
         private RegexWriter()
         {
             _intStack = new int[32];
-            _emitted = new int[32];
+            _emitted = new List<int>(32);
```
Comment (ViktorHofer) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: I think we should use a larger size here to avoid the cost of resizing for long patterns. E.g. 50 would better a fit and would only increase memory memory by 72 byte

**Comment by ViktorHofer** [comment]
> Perhaps it would make sense to share how to do this and how to run them, if it's easier than xunit.perf.

We already have a BenchmarkDotNet docu which described exactly how to test with 2.1 from master. What would you like to be further explained? The conversion from the xunit test to a BDN test was done manually by simply changing attributes and copying testdata code.

> Do you happen to have a ZIP file from PerfView from the perf test? Just would be interseting to look at. There are innumerable possible places to improve. Eg., should we construct Boyer Moore tables when the prefix or text are very small.

Not yet. I tried to use the VS perf feature but for some reason it doesn't include the src assembly correctly while recording the CPU. I will do what you suggest :)

> We should probably add another perf test to exercise RegexOptions.Compiled, and one to exercise caching.

For the cache benchmark see https://github.com/dotnet/corefx/issues/24425#issuecomment-361280655. @artkpv do you want to work on that?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 2]
```diff
@@ -299,7 +262,7 @@ private RegexCode RegexCodeFromRegexTree(RegexTree tree)
 
             anchors = RegexFCD.Anchors(tree);
 
-            return new RegexCode(_emitted, _stringtable, _trackcount, _caps, capsize, bmPrefix, fcPrefix, anchors, rtl);
+            return new RegexCode(_emitted.ToArray(), _stringtable, _trackcount, _caps, capsize, bmPrefix, fcPrefix, anchors, rtl);
```
Comment (ViktorHofer) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: ToArray copies the elements into a new array. It might make more sense to just return the List<int> as an IEnumerable<int> here.

**Comment by ViktorHofer** [comment]
cc @stephentoub for my two inline code comments

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 1 above)
Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: Whatever you do should be supported by measurements...

Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: > to avoid the cost of resizing for long patterns

Do we have any data on the distribution of pattern lengths?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: As an IEnumerable?  Doesn't RegexCode need to be able to index into it?  If anything I'd expect you'd just store the existing List as a List.  But as Dan said earlier, everything should be supported by measurements.

**Comment by cod7alex** [comment]
Maybe you can use ValueListBuilder here which was recently added for somewhat similar purpose.
https://github.com/dotnet/corefx/blob/0ff2a6f5dc55a6b82056ae4d29d4305f77e94689/src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: > Average is about 58 op codes.

What data set is this based on?

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: If there's a correlation between pattern length and opcode count you could possibly use a multiplier of some sort.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: Because of character sets and "multis" (string text) in most patterns, there is no strong correlation...

**Comment by ViktorHofer** [comment]
Results with latest commit:

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10.0.17093
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical cores and 2 physical cores
Frequency=2742188 Hz, Resolution=364.6723 ns, Timer=TSC
.NET Core SDK=2.1.100-preview-007326
  [Host]               : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  .NET Core 2.1 master : .NET Core 2.1.0-preview2-26130-06 (Framework 4.6.26205.05), 64bit RyuJIT
  .NET Core 2.1 regex  : .NET Core 2.1.0-preview2-26130-06 (Framework 4.6.26205.05), 64bit RyuJIT

Runtime=Core  Toolchain=OutOfProcessToolchain  

```
|     Method |                  Job |     Mean |     Error |    StdDev |    Gen 0 | Allocated |
|----------- |--------------------- |---------:|----------:|----------:|---------:|----------:|
| RegexMatch | .NET Core 2.1 master | 1.817 ms | 0.0360 ms | 0.0527 ms | 937.5000 |   1.88 MB |
| RegexMatch |  .NET Core 2.1 regex | 1.678 ms | 0.0324 ms | 0.0443 ms | 882.8125 |   1.77 MB |

**Comment by ViktorHofer** [comment]
Thanks @cod7alex for your suggestion to use the ValueListBuilder. It wasn't as easy as I thought because of all the constraints with ref structs but I believe I found a solution for that.

@stephentoub could you please glance over my latest commit. I'm slightly concerned that I'm misusing the ref struct concept (I'm building a big struct here just to be able to use spans and valuelistbuilder as a field...). thanks!

**Comment by danmosemsft** [comment]
Nit, its harder to review a change that straddles a commit changing code style. Ideally that would the the first or last commit or a different PR :)

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 3]
```diff
@@ -109,7 +132,7 @@ private int CurPos()
         /// </summary>
         private void PatchJump(int offset, int jumpDest)
         {
-            _emitted[offset + 1] = jumpDest;
+            Emitted[offset + 1] = jumpDest;
```
Comment (cod7alex) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: can`t builder just be passed as ref here?

**Comment by cod7alex** [comment]
@ViktorHofer I don`t think ValueListBuilder is suitable now to be stored as a field. Also there are no calls to `Dispose` as i can see which will lead to misuse of `ArrayPool.Shared`.

**Comment by ViktorHofer** [comment]
> Nit, its harder to review a change that straddles a commit changing code style. Ideally that would the the first or last commit or a different PR :)

You are right. As the commit wasn't really substantial I dropped it. I maybe will submit a PR to cleanup and pretty-format the source.

> Also there are no calls to Dispose as i can see which will lead to misuse of ArrayPool.Shared.

Fixed.

> @ViktorHofer I dont think ValueListBuilder is suitable now to be stored as a field.

I agree. But I will wait for what @stephentoub has to say about it.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: I tried that approach which resulted in all methods requiring 2-3 more parameters. Passing them around degrades the codes readability a lot. The initial design of this class was to have a static entry point `RegexWriter.Write` which then creates a private instance of the class just to make it easier to share fields. I'm not huge fan of that either as it involves allocation of an unnecessary object on the heap (it was a class before).

**Comment by stephentoub** [comment]
> I agree. But I will wait for what @stephentoub has to say about it.

What specifically are you asking about?  There are a ton of changes in that second commit that don't appear to be related to the ValueStringBuilder change, and it's hard to follow.

**Comment by ViktorHofer** [comment]
> What specifically are you asking about? There are a ton of changes in that second commit that don't appear to be related to the ValueStringBuilder change, and it's hard to follow.

The second commit only contains the changes that were necessary to use ValueStringBuilder. The diff is hard to follow here.

My question is if changing the data structure to a ref struct to be able to use Span and ValueListBuilder as fields should be avoided? The struct isn't passed anywhere around. Also please see my previous comment https://github.com/dotnet/corefx/pull/26877#discussion_r167560723

**Comment by ViktorHofer** [comment]
My results don't look that promising anymore. 

1. With ValueListBuilder the performance is the same as in master but the allocation is much less.
2. With List the performance is faster but the allocation is "slightly" higher.

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10.0.17093
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical cores and 2 physical cores
Frequency=2742188 Hz, Resolution=364.6723 ns, Timer=TSC
  [Host] : .NET Core ? (Framework 4.6.26205.05), 64bit RyuJIT

Job=.NET Core 2.1 regex  Runtime=Core  Toolchain=InProcessToolchain  

```
|     Method |     Mean |     Error |    StdDev |    Gen 0 | Allocated |
|----------- |---------:|----------:|----------:|---------:|----------:|
| RegexMatch (master) | 2.401 ms | 0.0542 ms | 0.1563 ms | 937.5000 |   1.88 MB |
| RegexMatch (list --> commit 1) | 2.216 ms | 0.0441 ms | 0.1266 ms | 941.4063 |   1.89 MB |
| RegexMatch (valuelistbuilder --> commit 1&2) | 2.403 ms | 0.0790 ms | 0.2108 ms | 882.8125 |   1.77 MB |

**Comment by ViktorHofer** [comment]
I just had a crazy idea how to get rid off the struct wrapper 🗡

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 4]
```diff
@@ -18,89 +18,118 @@
...
-        /// This is the push.
+        /// The top level RegexCode generator. It does a depth-first walk
+        /// through the tree and calls EmitFragment to emits code before
+        /// and after each child of an interior node, and at each leaf.
         /// </summary>
-        private void PushInt(int i)
+        public static RegexCode Write(RegexTree tree)
         {
-            if (_depth >= _intStack.Length)
+            Span<int> emittedSpan = stackalloc int[48];
```
Comment (joshfree) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: nit: can you replace the 48/32's with private const ints

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 5]
```diff
@@ -18,89 +18,118 @@
...
-        private int[] _intStack;
-        private int _depth;
-        private int[] _emitted;
-        private int _curpos;
-        private readonly Dictionary<string, int> _stringhash;
-        private readonly List<string> _stringtable;
-        private bool _counting;
-        private int _count;
-        private int _trackcount;
-        private Hashtable _caps;
```
Comment (joshfree) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: I think these changes will break someone who used RegexCompileToAssembly on desktop fx trying to then use that assembly in a .NET Core 2.1+ app

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: afaik we don't support that scenario. I was discussing that with Dan previously. And I'm not sure if RegexWriter is cached in the compiled assembly. I don't think so.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: Passing the ValueBuilders as ref degrades performance substantially because of the passing of refs (just tried that, costs doubled).

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs [hunk 6]
```diff
@@ -34,6 +36,13 @@ public void Append(T item)
             _pos = pos + 1;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public T Pop()
+        {
+            _pos--;
+            return _span[_pos];
+        }
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: Do we need to add this to ValueListBuilder?  Seems like it could just be a helper in the regex code.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs [hunk 7]
```diff
@@ -18,89 +18,118 @@
...
+        public static RegexCode Write(RegexTree tree)
         {
-            if (_depth >= _intStack.Length)
+            Span<int> emittedSpan = stackalloc int[48];
+            Span<int> intStackSpan = stackalloc int[32];
+            RegexWriter writer = new RegexWriter(emittedSpan, intStackSpan);
+
+            // construct sparse capnum mapping if some numbers are unused
+            int capsize;
+            if (tree._capnumlist == null || tree._captop == tree._capnumlist.Length)
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: Why did you manually inline RegexCodeFromRegexTree?  It makes it difficult to see any differences.  Was it necessary for some reason?

File: src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs (hunk 6 above)
Comment (ViktorHofer) [reply src/Common/src/CoreLib/System/Collections/Generic/ValueListBuilder.cs]: I agree, I think this should be a dedicated type in Regex.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: It wasn't really necessary, that might fall under over-engineering. I can switch back to what it was before.

**Comment by ViktorHofer** [comment]
This PR is now ready and I will merge it if no concerns are raised by EOD. Before I merge I will of course do another benchmarking round, just to be sure.

**Comment by stephentoub** [comment]
@ViktorHofer, now that System.Text.RegularExpressions is using stackalloc, you might want to look at https://github.com/dotnet/corefx/pull/26993 and do the same for regex, and see if there's any measurable impact on throughput.

**Comment by danmosemsft** [comment]
Should ResizableValueListBuilder be in Common, not S.T.RE? (With Pop() extracted into Regex code)

**Comment by stephentoub** [comment]
> Should ResizableValueListBuilder be in Common, not S.T.RE? (With Pop() extracted into Regex code)

ValueListBuilder is in common. I didn't actually mean to suggest that a copy of it be made; rather that S.T.RE just contain the pop method, e.g. a static method taking the VLE by ref as an argument. I'd prefer we not have multiple copies of essentially the same thing.

**Comment by ViktorHofer** [comment]
Thanks for the feedback. It seems I rushed completing this PR. I addressed feedback in https://github.com/dotnet/corefx/pull/27146

**Comment by artkpv** [comment]
>> We should probably add another perf test to exercise RegexOptions.Compiled, and one to exercise caching.

> For the cache benchmark see #24425 (comment). @artkpv do you want to work on that?

I'll check this. Still i'm slow as learning corefx interfals still

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs (hunk 5 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: @joshfree I want to verify, are you sure this breaks the `RegexCompileToAssembly` scenario? I just checked and the only places where the RegexWriter is used in that process is when RegexCode is generated. And only the capsize (int) and the caps (Hashtable) are written into IL from RegexCode.

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: https://referencesource.microsoft.com/#System/regex/system/text/regularexpressions/RegexCompiler.cs,3136

Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: sure, will do. I don't think we broke anything here but if we really want to stay compatible with netfx CompileToAssembly generated assemblies we might restrict ourselves in modernizing the existing engine.

Comment (danmosemsft) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: Ideally we should have a test checked in that runs a compiled regex (emitted by desktop). We have other tests that have small projects to build ancilliary test binaries.

Comment (joshfree) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs]: if need be, you can persist the test binary asset(s) at https://github.com/dotnet/corefx-testdata
